api/saml/configuration/model.py

```diff
@@ -0,0 +1,604 @@
+    ConfigurationAttributeType,
+    ConfigurationFactory,
+    ConfigurationGrouping,
+    ConfigurationMetadata,
+    ConfigurationOption,
+)
+
+
+class SAMLConfigurationError(BaseError):
+    """Raised in the case of any configuration errors."""
+
+
+class SAMLConfiguration(ConfigurationGrouping):
+    """Contains SP and IdP settings."""
+
+    service_provider_xml_metadata = ConfigurationMetadata(
+        key="sp_xml_metadata",
+        label=_("Service Provider's XML Metadata"),
+        description=_(
+            "SAML metadata of the Circulation Manager's Service Provider in an XML format. "
+            "MUST contain exactly one SPSSODescriptor tag with at least one "
+            "AssertionConsumerService tag with Binding attribute set to "
+            "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST."
+        ),
+        type=ConfigurationAttributeType.TEXTAREA,
+        required=True,
+    )
+
+    service_provider_private_key = ConfigurationMetadata(
+        key="sp_private_key",
+        label=_("Service Provider's Private Key"),
+        description=_("Private key used for encrypting SAML requests."),
+        type=ConfigurationAttributeType.TEXTAREA,
+        required=False,
+    )
+
+    federated_identity_provider_entity_ids = ConfigurationMetadata(
+        key="saml_federated_idp_entity_ids",
+        label=_("List of Federated IdPs"),
+        description=_(
+            "List of federated (for example, from InCommon Federation) IdPs supported by this authentication provider. "
+            "Try to type the name of the IdP to find it in the list."
+        ),
+        type=ConfigurationAttributeType.MENU,
+        required=False,
+        options=[],
+        default=[],
+        format="narrow",
+    )
+
+    non_federated_identity_provider_xml_metadata = ConfigurationMetadata(
+        key="idp_xml_metadata",
+        label=_("Identity Provider's XML metadata"),
+        description=_(
+            "SAML metadata of Identity Providers in an XML format. "
+            "MAY contain multiple IDPSSODescriptor tags but each of them MUST contain "
+            "at least one SingleSignOnService tag with Binding attribute set to "
+            "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect."
+        ),
+        type=ConfigurationAttributeType.TEXTAREA,
+        required=False,
+    )
+
+    filter_expression = ConfigurationMetadata(
+        key="saml_filter_expression",
+        label=_("Filter Expression"),
+        description=_(
+            "Python expression used for filtering out patrons by their SAML attributes."
+            "<br>"
+            "<br>"
+            'For example, if you want to authenticate using SAML only patrons having "eresources" '
+            'as the value of their "eduPersonEntitlement" then you need to use the following expression:'
+            "<br>"
+            '<pre>subject.attribute_statement.attributes["eduPersonEntitlement"].values[0] == "eresources"</pre>'
```

Comment:
So this answers the question I asked initially -- we're denying patrons access based on the results of the 'filter' -- but now I have another question. Is this a DSL or are we asking people to write Python code here? That seems really bad, both in terms of usability and security.